### PR TITLE
Spitting LGFS fee pages

### PR DESCRIPTION
**Pull request by lostie**
For the time being, until #2241 gets merge the actual changes can be found [here](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/compare/spitting-agfs-fee-pages...spitting-lgfs-fee-pages?expand=1)

#### What
Split the LGFS fees page into different fee pages for each fee types, according to the most updated LGFS claim submission flows.

#### Why
Result of the user research made to simplify the claim submission.
 
#### How

- Each fee type (**basic and/or fixed**, **graduated**, **interim**, **miscellaneous**, **warrant**) and, where applicable, **disbursements**, **travel expenses**, **supporting evidence** and **additional information** now have their own page.
- Each claim type model now has a stage management that controls which stages are required for the claim.

Also, some other improvements:

- Rather than render every single page in the claim submission flow (which was causing re-submission of the current form on page reload), the user will be redirected to the page for the next stage (which does not have any issues with page reloading and it's fairly conventional to do anyway). The only time the current form page will be rendered (not redirected) after the form submission, is when the current form contains errors.
- The previous improvement also helped to introduce _stepable_ forms for the claim submission, meaning it will be easier in the future to navigate backwards and forwards through the different stages/steps and even edit a specific stage/step of the claim without having to go through the whole set of stages/steps.
- Improved validation of claim certifications, mainly to ensure the date of certification is after the claim was created and is not in the future.
- Removed unused validations (certifications, messages, assessments, redeterminations) for `has_many/has_one` relations on the claim validators, as they have their own validation, not using this ones.

#### Ticket

[Pivotal Tracker](https://www.pivotaltracker.com/story/show/154964797)

### CODE REVIEW

File: app/models/claim/base_claim.rb [hunk 1]
```diff
@@ -389,25 +393,18 @@ def disabled_for_transition?
...
     end
 
-    def current_step_index
-      submission_stages.index(current_step)
+    def submission_current_flow
+      return submission_stages if from_api?
+      submission_stages.path_until(form_step)
     end
 
-    def current_step_required?
```
Comment (lostie) [inline app/models/claim/base_claim.rb:456]: 🏌️ Got rid of this completely, as any conditions are now managed through the submission _stage manager_.

File: app/validators/base_sub_model_validator.rb [hunk 2]
```diff
@@ -35,6 +35,7 @@ def validate_has_one_associations(record)
   def validate_collection_for(record, association_name)
     collection = record.__send__(association_name)
     collection.each_with_index do |associated_record, i|
+      next if associated_record.marked_for_destruction?
```
Comment (lostie) [inline app/validators/base_sub_model_validator.rb]: ⚠️  This is one of the new **code climate** _offended_ files due to this change?! 😕

File: app/validators/claim/advocate_claim_sub_model_validator.rb [hunk 3]
```diff
@@ -4,29 +4,20 @@ def has_one_association_names_for_steps
...
-        :defendants
-      ],
+      defendants: %i[defendants],
       offence_details: [],
-      fees: %i[
-        basic_fees
-        misc_fees
-        fixed_fees
-        expenses
-        messages
```
Comment (lostie) [inline app/validators/claim/advocate_claim_sub_model_validator.rb:26]: 📝 ⚠️  I've tested this several times, tracked which controllers and validators where taking care of `messages`, `redeterminations` and could not see this validator being used at all for those models.

File: app/validators/claim/interim_claim_sub_model_validator.rb [hunk 4]
```diff
@@ -4,11 +4,7 @@ def has_one_association_names_for_steps
       case_details: [],
       defendants: [],
       offence_details: [],
-      fees: %i[
-        interim_fee
-        assessment
```
Comment (lostie) [inline app/validators/claim/interim_claim_sub_model_validator.rb:9]: [Ditto](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/2252/files#r170615857)

File: app/validators/claim/interim_claim_sub_model_validator.rb [hunk 5]
```diff
@@ -17,12 +13,9 @@ def has_many_association_names_for_steps
       case_details: [],
       defendants: %i[defendants],
       offence_details: [],
-      fees: %i[
-        disbursements
-        messages
```
Comment (lostie) [inline app/validators/claim/interim_claim_sub_model_validator.rb:22]: [Ditto](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/2252/files#r170615857)

File: app/validators/claim/litigator_claim_sub_model_validator.rb [hunk 6]
```diff
@@ -4,13 +4,9 @@ def has_one_association_names_for_steps
       case_details: [],
       defendants: [],
       offence_details: [],
-      fees: %i[
-        graduated_fee
-        fixed_fee
-        warrant_fee
-        assessment
```
Comment (lostie) [inline app/validators/claim/litigator_claim_sub_model_validator.rb:11]: [Ditto](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/2252/files#r170615857)

File: app/validators/claim/litigator_claim_sub_model_validator.rb [hunk 7]
```diff
@@ -19,14 +15,10 @@ def has_many_association_names_for_steps
       case_details: [],
       defendants: %i[defendants],
       offence_details: [],
-      fees: %i[
-        misc_fees
-        disbursements
-        expenses
-        messages
```
Comment (lostie) [inline app/validators/claim/litigator_claim_sub_model_validator.rb:26]: [Ditto](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/2252/files#r170615857)

File: app/validators/claim/transfer_claim_sub_model_validator.rb [hunk 8]
```diff
@@ -5,11 +5,7 @@ def has_one_association_names_for_steps
       case_details: [],
       defendants: [],
       offence_details: [],
-      fees: %i[
-        transfer_fee
-        assessment
```
Comment (lostie) [inline app/validators/claim/transfer_claim_sub_model_validator.rb:10]: [Ditto](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/2252/files#r170615857)

File: app/validators/claim/transfer_claim_sub_model_validator.rb [hunk 9]
```diff
@@ -19,14 +15,10 @@ def has_many_association_names_for_steps
       case_details: [],
       defendants: %i[defendants],
       offence_details: [],
-      fees: %i[
-        misc_fees
-        disbursements
-        expenses
-        messages
```
Comment (lostie) [inline app/validators/claim/transfer_claim_sub_model_validator.rb:26]: [Ditto](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/2252/files#r170615857)

File: features/claims/litigator/interim_trial_warrant_claim_draft_submit.feature [hunk 10]
```diff
@@ -0,0 +1,67 @@
+@javascript
```
Comment (lostie) [inline features/claims/litigator/interim_trial_warrant_claim_draft_submit.feature:1]: 📝  This feature was added to cover the path where **warrant fees** are required. The existent _feature_ covers the path where they're not.

**Review by colinbruce — APPROVED** [review]
Looks fine from a code perspective.
👍 
I'd appreciate a 10 minute run through of how the new stage management works.

File: spec/validators/claim/interim_claim_sub_model_validator_spec.rb [hunk 11]
```diff
@@ -2,20 +2,22 @@
 require_relative 'shared_examples_for_step_validators'
 
 RSpec.describe Claim::InterimClaimSubModelValidator, type: :validator do
-  let(:claim) { FactoryBot.create :interim_claim }
+  let(:claim) { create(:interim_claim) }
```
Comment (jsugarman) [inline spec/validators/claim/interim_claim_sub_model_validator_spec.rb:5]: 👍

File: spec/validators/claim/advocate_claim_validator_spec.rb [hunk 12]
```diff
@@ -294,9 +293,10 @@
       ],
       [],
       %i[offence],
-      %i[
-        total
-        defendant_uplifts
-      ]
+      %i[total defendant_uplifts],
+      %i[total],
```
Comment (jsugarman) [inline spec/validators/claim/advocate_claim_validator_spec.rb]: does this validate total > 0 for each fee page?

File: spec/validators/claim/litigator_claim_sub_model_validator_spec.rb [hunk 13]
```diff
@@ -9,13 +9,20 @@
...
       [],
-      [:defendants],
+      %i[defendants],
       [],
-      [:misc_fees, :disbursements, :expenses, :messages, :redeterminations, :documents]
+      %i[misc_fees],
+      %i[disbursements],
+      %i[expenses],
+      %i[documents]
+      # [:misc_fees, :disbursements, :expenses, :messages, :redeterminations, :documents]
```
Comment (jsugarman) [inline spec/validators/claim/litigator_claim_sub_model_validator_spec.rb]: do you want to keep this here?

File: spec/support/validation_helpers.rb [hunk 14]
```diff
@@ -21,7 +15,7 @@ def create_and_submit_claim(claim)
   end
 
   def should_error_with(record, field, message)
-    expect(record.valid?).to be false
+    expect(record).not_to be_valid
```
Comment (jsugarman) [inline spec/support/validation_helpers.rb:18]: 👍

File: app/models/claim/advocate_claim.rb [hunk 15]
```diff
@@ -83,10 +83,92 @@ class AdvocateClaim < BaseClaim
...
+        transitions: [
+          { to_stage: :defendants }
+        ]
+      },
+      {
+        name: :defendants,
+        transitions: [
+          {
+            to_stage: :offence_details,
+            condition: ->(claim) { !claim.fixed_fee_case? }
```
Comment (jsugarman) [inline app/models/claim/advocate_claim.rb:101]: I like this - surprised it does not raise a dynamic constant assignment rubocop violation tho'

File: app/models/claim/interim_claim.rb [hunk 16]
```diff
@@ -76,8 +76,65 @@ class InterimClaim < BaseClaim
...
+        ]
+      },
+      {
+        name: :offence_details,
+        transitions: [
+          { to_stage: :interim_fees }
+        ]
+      },
+      {
+        name: :interim_fees,
```
Comment (jsugarman) [inline app/models/claim/interim_claim.rb:99]: not a blocker - but there is only one interim_fee on an interim claim (`has_one :interim_fee`)

File: app/models/claim/litigator_claim.rb [hunk 17]
```diff
@@ -82,8 +82,96 @@ class LitigatorClaim < BaseClaim
...
+        ]
+      },
+      {
+        name: :offence_details,
+        transitions: [
+          { to_stage: :graduated_fees }
+        ]
+      },
+      {
+        name: :fixed_fees,
```
Comment (jsugarman) [inline app/models/claim/litigator_claim.rb:112]: not a blocker - but there is only one fixed_fee on an interim claim (`has_one :fixed_fee`)

this is different on AGFS which has_many :fixed_fees

File: app/models/claim/litigator_claim.rb [hunk 18]
```diff
@@ -82,8 +82,96 @@ class LitigatorClaim < BaseClaim
...
+          {
+            to_stage: :fixed_fees,
+            condition: ->(claim) { claim.fixed_fee_case? }
+          }
+        ]
+      },
+      {
+        name: :offence_details,
+        transitions: [
+          { to_stage: :graduated_fees }
```
Comment (jsugarman) [inline app/models/claim/litigator_claim.rb:108]: not a blocker - but there is only one graduated_fee on an interim claim (`has_one :graduated_fee`)

File: app/models/claim/litigator_claim.rb [hunk 19]
```diff
@@ -82,8 +82,96 @@ class LitigatorClaim < BaseClaim
...
+            condition: ->(claim) { claim.fixed_fee_case? }
+          },
+          {
+            to_stage: :warrant_fees,
+            condition: ->(claim) { !claim.fixed_fee_case? }
+          }
+        ]
+      },
+      {
+        name: :warrant_fees,
```
Comment (jsugarman) [inline app/models/claim/litigator_claim.rb]: ditto on the plural

File: app/models/claim/transfer_claim.rb [hunk 20]
```diff
@@ -93,8 +93,76 @@ class TransferClaim < BaseClaim
...
+        ]
+      },
+      {
+        name: :offence_details,
+        transitions: [
+          { to_stage: :transfer_fees }
+        ]
+      },
+      {
+        name: :transfer_fees,
```
Comment (jsugarman) [inline app/models/claim/transfer_claim.rb:122]: ditto on the plural

File: app/validators/claim/transfer_claim_sub_model_validator.rb [hunk 21]
```diff
@@ -5,11 +5,12 @@ def has_one_association_names_for_steps
       case_details: [],
       defendants: [],
       offence_details: [],
-      fees: %i[
-        transfer_fee
-        assessment
-        certification
-      ]
+      transfer_fees: %i[transfer_fee]
```
Comment (jsugarman) [inline app/validators/claim/transfer_claim_sub_model_validator.rb:8]: related to my singular plural comments - does it make a difference if you use the same, singular, name ` transfer_fee: %i[transfer_fee]`

**Review by jsugarman — APPROVED** [review]
👍 like the submission_stage logic. some non-blocking comments and obviously sidebar issues prevent merge

**Pull request merged**